lib/isorender/dom-shims.js

```diff
@@ -27,28 +27,50 @@ import requestAnimationFrame from 'raf';
 // is already there
 global.requestAnimationFrame = global.requestAnimationFrame || requestAnimationFrame;
 
+function walkDomTree(root, callback) {
+  // basic breadth-first tree traversal (non-recursive)
+  const breadthQueue = [root];
+  while (breadthQueue.length > 0) {
+    const node = breadthQueue.shift();
+    callback(node);
+    for (const child of Array.from(node.childNodes || [])) {
+      breadthQueue.push(child);
+    }
+    if (node.shadowRoot) {
+      for (const child of Array.from(node.shadowRoot.childNodes || [])) {
+        breadthQueue.push(child);
+      }
+    }
```

Comment:
Hmm it might be better to run the connected callback _after_ adding the children to the breadth queue so that children appended during `connectedCallback()` aren't also traversed. I'll give this a go next week.

Reply:
Never mind, that doesn't work. Instead, I've just added an if-check to guard against redundant lifecycle calls.